language/move-vm/runtime/src/runtime.rs

```diff
@@ -111,58 +113,127 @@ impl VMRuntime {
+        let mut kinfos = vec![];
+        for ty in tys {
+            match self.loader.type_to_type_layout(ty) {
+                Ok(layout) => layouts.push(layout),
+                Err(err) => {
+                    error!("[VM] failed to get layout from type");
+                    return Err(err);
+                }
+            }
+
+            match self.loader.type_to_kind_info(ty) {
+                Ok(kinfo) => kinfos.push(kinfo),
+                Err(err) => {
+                    error!("[VM] failed to get kind info from type");
+                    return Err(err);
+                }
+            }
+        }
+
+        // Try to deserialize the arguments according to the layouts.
+        let mut vals = vec![];
+        for ((layout, kinfo), arg) in layouts.iter().zip(kinfos.iter()).zip(args.iter()) {
+            match Value::simple_deserialize(arg, kinfo, layout) {
+                Some(val) => vals.push(val),
+                None => {
+                    warn!("[VM] failed to deserialize argument");
+                    return Err(PartialVMError::new(
+                        StatusCode::FAILED_TO_DESERIALIZE_RESOURCE,
+                    ));
+                }
+            }
+        }
+
+        Ok(vals)
+    }
+
+    fn create_signers_and_arguments(
         &self,
-        script: Vec<u8>,
-        ty_args: Vec<TypeTag>,
-        mut args: Vec<Value>,
+        tys: &[Type],
         senders: Vec<AccountAddress>,
-        data_store: &mut impl DataStore,
-        cost_strategy: &mut CostStrategy,
-        log_context: &impl LogContext,
-    ) -> VMResult<()> {
+        args: Vec<Vec<u8>>,
+    ) -> PartialVMResult<Vec<Value>> {
         // signer helper closure
-        fn is_signer_reference(s: &SignatureToken) -> bool {
-            use SignatureToken as S;
+        fn is_signer_reference(s: &Type) -> bool {
             match s {
-                S::Reference(inner) => matches!(&**inner, S::Signer),
+                Type::Reference(ty) => matches!(&**ty, Type::Signer),
                 _ => false,
             }
         }
 
-        // load the script, perform verification
-        let (main, type_params) =
-            self.loader
-                .load_script(&script, &ty_args, data_store, log_context)?;
-
-        // Build the arguments list for the main and check the arguments are of restricted types.
+        // Build the arguments list and check the arguments are of restricted types.
         // Signers are built up from left-to-right. Either all signer arguments are used, or no
         // signer arguments can be be used by a script.
-        let parameters = &main.parameters().0;
-        let has_signer_parameters = parameters.get(0).map_or(false, is_signer_reference);
-        let mut signers_and_args = if has_signer_parameters {
-            if parameters.len() != args.len() + senders.len() {
+        let has_signer_parameters = tys.get(0).map_or(false, is_signer_reference);
```

Comment:
Pre-existing, but: will this do the right thing for (e.g.) `main(i: u64, a: &signer)`? It seems funky to only look at the first type.

Reply:
> will this do the right thing for (e.g.) main(i: u64, a: &signer)?

No, since it is currently required that signer arguments come before the regular ones, but I think we can definitely change it if we wish. Thoughts @tnowacki @tzakian?

Reply:
Ah, by "right thing" I meant reject the code. I am totally ok with requiring the signer arguments to come first (and think it should stay that way).

Reply:
Oh I see. My understanding is that the current code will reject them, but the error messages may look weird and the errors code may not be correct. I'll do another round of review on errors after I fix our existing tests, and will write additional tests to cover inputs we didn't allow in the past.